### PR TITLE
Getting property values: fix potential cache key conflict, improve speed

### PR DESCRIPTION
**Pull request by lucaspimentel**
TL;DR
- replace uses of `Type.FullName` with `Type.AssemblyQualifiedName` in `MethodBuilder` and `ObjectExtensions`
- use the `PropertyFetcher` helper instead of emitting IL in `ObjectExtensions.TryGetPropertyValue()`
- replace the use of reflection to access a field in the `StackExchange.Redis` instrumentation

This PR fixes an issue with the cache key used to store dynamically emitted methods. A conflict can occur when we encounter multiple classes with the same full name. We would reuse the cached dynamic method, and the emitted IL would try to cast one class into the other, but they are incompatible classes and causes an `InvalidCastException`. Using the assembly qualified name for types fixes this since includes assembly name and version to generate different cache keys.

This PR also replaces the use of dynamic methods with `PropertyFetcher` (added in #611) when getting property values. Using `PropertyFetcher` is much faster than emitting IL and doesn't allocate any memory for each fetch. These are benchmark results from testing a very simple scenario using `ObjectExtensions.GetProperty()` and `PropertyFetcher.Fetch()` to the get value of a property:

```
|                         Method |       Runtime |        Mean |     Error |    StdDev |      Median | Allocated |
|------------------------------- |-------------- |------------:|----------:|----------:|------------:|----------:|
| ObjectExtensions.GetProperty() | .NET 4.6.1    | 1,157.07 ns | 18.849 ns | 15.740 ns | 1,151.96 ns |    2392 B |
| ObjectExtensions.GetProperty() | .NET Core 2.1 | 1,141.81 ns | 11.196 ns | 10.473 ns | 1,139.89 ns |    2256 B |
| ObjectExtensions.GetProperty() | .NET Core 3.1 | 1,082.25 ns |  8.152 ns |  7.625 ns | 1,082.74 ns |    2240 B |
|------------------------------- |-------------- |------------:|----------:|----------:|------------:|----------:|
| PropertyFetcher.Fetch()        | .NET 4.6.1    |    23.31 ns |  0.317 ns |  0.297 ns |    23.32 ns |         - |
| PropertyFetcher.Fetch()        | .NET Core 2.1 |    20.61 ns |  0.207 ns |  0.193 ns |    20.54 ns |         - |
| PropertyFetcher.Fetch()        | .NET Core 3.1 |    17.99 ns |  0.392 ns |  1.060 ns |    17.58 ns |         - |
```

Bonus change: reduce number of steps in `integration-tests.yaml` pipeline

To do:
- [x] add reproduction project
- [x] test both versions of Sigil
- [x] test on .NET Framework
- [x] test on .NET Core
- [x] run regression test in CI

### CODE REVIEW

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/SmokeTests/StackExchangeRedisAssemblyConflictLegacyProjectSmokeTest.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+namespace Datadog.Trace.ClrProfiler.IntegrationTests.SmokeTests
+{
+    public class StackExchangeRedisAssemblyConflictLegacyProjectSmokeTest : SmokeTestBase
+    {
+        public StackExchangeRedisAssemblyConflictLegacyProjectSmokeTest(ITestOutputHelper output)
+            : base(output, "StackExchange.Redis.AssemblyConflict.LegacyProject", maxTestRunSeconds: 15)
+        {
+        }
+
+        [TargetFrameworkVersionsFact("net452;net461")]
```
Comment (bobuva) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/SmokeTests/StackExchangeRedisAssemblyConflictLegacyProjectSmokeTest.cs]: Nice, didn't know that was a thing.

Comment (lucaspimentel) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/SmokeTests/StackExchangeRedisAssemblyConflictLegacyProjectSmokeTest.cs]: @zacharycmontoya wrote it! It's a specialized `[Fact]` that sets the `Skip` property if the target framework doesn't match.

File: reproductions/StackExchange.Redis.AssemblyConflict.LegacyProject/Program.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                Console.WriteLine($"Datadog.Trace.dll path: {tracerAssemblyLocation}");
+                Console.WriteLine($"Datadog.Trace.ClrProfiler.Managed.dll path: {clrProfilerAssemblyLocation}");
+                Console.WriteLine($"Sigil.dll path: {sigilAssemblyLocation}");
+                Console.WriteLine();
+            }
+        }
+
+        private static void RunTest()
+        {
+            const string configuration = "127.0.0.1:6389";
```
Comment (lucaspimentel) [inline reproductions/StackExchange.Redis.AssemblyConflict.LegacyProject/Program.cs]: Fix this for CI.

File: reproductions/StackExchange.Redis.AssemblyConflict.SdkProject/Program.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                Console.WriteLine($"Datadog.Trace.dll path: {tracerAssemblyLocation}");
+                Console.WriteLine($"Datadog.Trace.ClrProfiler.Managed.dll path: {clrProfilerAssemblyLocation}");
+                Console.WriteLine($"Sigil.dll path: {sigilAssemblyLocation}");
+                Console.WriteLine();
+            }
+        }
+
+        private static void RunTest()
+        {
+            const string configuration = "127.0.0.1:6389";
```
Comment (lucaspimentel) [inline reproductions/StackExchange.Redis.AssemblyConflict.SdkProject/Program.cs]: Fix this for CI.

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/SmokeTests/StackExchangeRedisAssemblyConflictLegacyProjectSmokeTest.cs (hunk 1 above)
Comment (bobuva) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/SmokeTests/StackExchangeRedisAssemblyConflictLegacyProjectSmokeTest.cs]: 💯 @zacharycmontoya 👍

File: reproductions/StackExchange.Redis.AssemblyConflict.SdkProject/Program.cs (hunk 3 above)
Comment (lucaspimentel) [reply reproductions/StackExchange.Redis.AssemblyConflict.SdkProject/Program.cs]: Fixed in 3dbcb76bda7136f8a05ace9221e04e6eb03d4e05

File: reproductions/StackExchange.Redis.AssemblyConflict.LegacyProject/Program.cs (hunk 2 above)
Comment (lucaspimentel) [reply reproductions/StackExchange.Redis.AssemblyConflict.LegacyProject/Program.cs]: Fixed in 3dbcb76bda7136f8a05ace9221e04e6eb03d4e05

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/SmokeTests/StackExchangeRedisAssemblyConflictLegacyProjectSmokeTest.cs (hunk 1 above)
Comment (lucaspimentel) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/SmokeTests/StackExchangeRedisAssemblyConflictLegacyProjectSmokeTest.cs]: Oops, I meant to include a link to [`TargetFrameworkVersionsFact`](https://github.com/DataDog/dd-trace-dotnet/blob/58908e43b3e1d5da3bcd19499709da6b2b259a83/test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/TargetFrameworkVersionsFact.cs).

File: src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs [hunk 4]
```diff
@@ -34,8 +34,8 @@ internal sealed class AspNetCoreDiagnosticObserver : DiagnosticObserver
         private static readonly PropertyFetcher HttpRequestInStopHttpContextFetcher = new PropertyFetcher("HttpContext");
         private static readonly PropertyFetcher UnhandledExceptionHttpContextFetcher = new PropertyFetcher("HttpContext");
         private static readonly PropertyFetcher UnhandledExceptionExceptionFetcher = new PropertyFetcher("Exception");
-        private static readonly PropertyFetcher BeforeActionHttpContextFetcher = new PropertyFetcher("HttpContext");
-        private static readonly PropertyFetcher BeforeActionActionDescriptorFetcher = new PropertyFetcher("ActionDescriptor");
+        private static readonly PropertyFetcher BeforeActionHttpContextFetcher = new PropertyFetcher("httpContext");
+        private static readonly PropertyFetcher BeforeActionActionDescriptorFetcher = new PropertyFetcher("actionDescriptor");
```
Comment (colin-higgins) [inline src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs]: What's the deal here? Was this a bug? Did this hit privates instead of publics or something?

Comment (lucaspimentel) [reply src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs]: The property names are actually lower case here. I just wanted them to match, but it doesn't make a real different since the code in case-insensitive. I noticed this because when I was making [this change](https://github.com/DataDog/dd-trace-dotnet/pull/642/files#diff-03ea678402f5d547eb5d427be0c0c583R35) I inadvertently make the code case-sensitive and the ASP.NET Core tests failed.

Comment (lucaspimentel) [reply src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs]: [Here's an example](https://github.com/dotnet/aspnetcore/blob/474d03c071867dacb40dbdef993d6714b276f558/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L229-L239) of how ASP.NET Core uses an anonymous classs to pass values into diagnostic listeners.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/StackExchange.Redis/StackExchangeRedisHelper.cs [hunk 5]
```diff
@@ -72,17 +71,7 @@ public static string GetRawCommand(object multiplexer, object message)
...
-            {
-                var fi = obj.GetType().GetField("multiplexer", BindingFlags.NonPublic | BindingFlags.Instance);
-                multiplexer = fi.GetValue(obj);
-            }
-            catch
-            {
-            }
-
-            return multiplexer;
+            return obj.GetField("multiplexer").GetValueOrDefault();
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/StackExchange.Redis/StackExchangeRedisHelper.cs]: Good find!
